### PR TITLE
[Ready for Review] Restructure to standalone sections

### PR DESCRIPTION
**Pull request by masonlr**
### Summary

This is a re-work of PR #575. The parent issue is #468.

### List of changes proposed in this PR (pull-request)

* Move content into sections (rather than single file)

### What should a reviewer concentrate their feedback on?
- [ ] Whether the table of contents is correct
- [ ] Whether the content has been divided into appropriate sections

### Acknowledging contributors

- [x] All contributors to this pull request are already named in the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file.

### CODE REVIEW

**Comment by masonlr** [comment]
Note, this PR incorporates the edits made to `book/content/reproducible_environments/*` by @alexmorley in be9627a2ff93ce3a5b215b935a84cec9100b5625

**Comment by masonlr** [comment]
We've lost the figures in the binder section. Will take a look at that now.

**Comment by masonlr** [comment]
Moved the images to within the `content` dir as per #375 discussion.

**Comment by rainsworth** [comment]
Thank you @masonlr! 👏  I'm going to run through the other PRs to make sure that there isn't any overlap in restructuring currently before reviewing this one.

**Comment by kkmann** [comment]
yipeeeeh! can be merged, right?

**Comment by masonlr** [comment]
Up to @rainsworth now.

File: book/content/reproducible_environments/06/containers.md [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+
+> Prerequisites: At present, Singularity only runs on linux systems (for example Ubuntu). If you use, macOS, [Singularity Desktop for macOS](https://www.sylabs.io/singularity-desktop-macos/) is in "Alpha Preview" stage.
+
+A major drawback of Docker for reproducible research is that it is not intended as a user-space application but as a tool for server administrators. As such it requires root access to operate. There is, however, no reason why the execution of an analysis should require root access for the user. This is especially important when computations are conducted on shared resource like HPC systems where users will never have root access.
+
+The [singularity](https://www.sylabs.io/) container software was introduced to address exactly this issue. Singularity was created with HPC sytems and reproducible research in mind
+(see [this](https://www.youtube.com/watch?v=DA87Ba2dpNM) video]. It does not require root access to run (only to build container _images_!) and thus enables HPC users to locally build container images before running analyses, e.g., on a high-performance cluster. As an added benefit, this makes it possible to use almost any software on an
```
Comment (rainsworth) [inline book/content/reproducible_environments/06/containers.md]:
```suggestion
(see [this](https://www.youtube.com/watch?v=DA87Ba2dpNM) video). It does not require root access to run (only to build container _images_!) and thus enables HPC users to locally build container images before running analyses, e.g., on a high-performance cluster. As an added benefit, this makes it possible to use almost any software on an
```

File: book/content/reproducible_environments/06/containers.md [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+
+Singularity can be used to run Docker images or extend them by building new images based on docker containers as base layer. For instance, we could use singularity to spin up a vanilla ubuntu container and getting a shell in it using the ubuntu docker image via
+
+```
+singularity shell docker://ubuntu
+```
+
+(type `exit` to leave the interactive shell again).
+
+Just as docker images are build using `Dockerfile` files, singularity containers
```
Comment (rainsworth) [inline book/content/reproducible_environments/06/containers.md]:
```suggestion
Just as docker images are built using `Dockerfile` files, singularity containers
```

File: book/content/reproducible_environments/06/containers.md [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+are build from singularity definition files.
```
Comment (rainsworth) [inline book/content/reproducible_environments/06/containers.md]:
```suggestion
are built from singularity definition files.
```

File: book/content/reproducible_environments/06/containers.md [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+The process and syntax is similar to docker files but there are subtle
+differences.
+As a minimal working example, we can build a 'lolcow' container based on the
+official ubuntu docker container image.
+Put the following in a `lolcow.def` file
+(based on the [Singularity documentation](https://www.sylabs.io/guides/3.2/user-guide/build_a_container.html))
```
Comment (rainsworth) [inline book/content/reproducible_environments/06/containers.md]:
```suggestion
(based on the [Singularity documentation](https://www.sylabs.io/guides/3.2/user-guide/build_a_container.html)):
```

File: book/content/reproducible_environments/06/containers.md [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+    fortune | cowsay | lolcat
+```
+
+This 'recipe' uses a docker image as basis (here: ubuntu) installs a few
+apt packages, modifies a few environment variables, and specifies the
+runscript (which is executed using the `singularity run` command).
+Details on the singularity definition file format can be found in the official
+[documentation](https://www.sylabs.io/docs/).
+
+A container image can then be build (requiring root!) via
```
Comment (rainsworth) [inline book/content/reproducible_environments/06/containers.md]:
```suggestion
A container image can then be built (requiring root!) via
```

File: book/content/reproducible_environments/06/containers.md [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+\ friend in trouble.                /
+-----------------------------------
+       \   ^__^
+        \  (oo)\_______
+           (__)\       )\/\
+               ||----w |
+               ||     ||
+```
+
+Being HPC compatible, singularity containers are also supported by wide range
```
Comment (rainsworth) [inline book/content/reproducible_environments/06/containers.md]:
```suggestion
Being HPC compatible, singularity containers are also supported by a wide range
```

File: book/content/reproducible_environments/06/containers.md [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+of workflow management tools.
+E.g. both [snakemake](https://snakemake.readthedocs.io/en/stable/) and
```
Comment (rainsworth) [inline book/content/reproducible_environments/06/containers.md]:
```suggestion
For example, both [snakemake](https://snakemake.readthedocs.io/en/stable/) and
```

File: book/content/reproducible_environments/06/containers.md [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+[nextflow](https://www.nextflow.io/docs/latest/singularity.html) support
+job-specific singularity containers.
+This makes singularity containers uniquely suited for parallelizing workflows
+on HPC systems using the
+widely used [slurm](https://slurm.schedmd.com/documentation.html) workload manager.
+Using singularity containers and snakemake/nextflow is therefore a way of
+scaling reproducibility to massive scale and - as an added benefit -
+brining workflows from a desktop machine to an HPC system no longer requires
```
Comment (rainsworth) [inline book/content/reproducible_environments/06/containers.md]:
```suggestion
bringing workflows from a desktop machine to an HPC system no longer requires
```

**Comment by rainsworth** [comment]
Thanks @masonlr and @kkmann! I just had a few minor comments on typos in the added Singularity content.

I will go ahead and approve this because the chapter looks great. Can I just ask that you please put each new sentence on a new line in the Singularity section? (I know the rest of the chapter doesn't follow this style guideline either, but ignore that for now) - there are just some sentences that are split over several lines in the Singularity section so it would make reviewing easier going forward.

Thank you so much for your contributions to this chapter!

File: book/content/reproducible_environments/reproducible_environments.md [hunk 9]
```diff
@@ -2,81 +2,84 @@
...
+  - [What is Docker?](./06/containers#What_is_Docker)
+  - [Installing Docker](./06/containers#Installing_Docker)
+  - [Key commands](./06/containers#Key_commands)
+  - [Writing Dockerfiles](./06/containers#Writing_Dockerfiles)
+    - [WORKDIR](./06/containers#WORKDIR)
+    - [Other commands](./06/containers#Other_commands)
+  - [Building images and .dockerignore files](./06/containers#Building_images_and_dockerignore_files)
+  - [Sharing images](./06/containers#Sharing_images)
+  - [Copying files to and from containers](./06/containers#Copying_files_to_and_from_containers)
+  - [Volumes](./06/containers#Volumes)
```
Comment (rainsworth) [inline book/content/reproducible_environments/reproducible_environments.md]: Add your singularity section?

Comment (masonlr) [reply book/content/reproducible_environments/reproducible_environments.md]: Done.

**Comment by masonlr** [comment]
> Can I just ask that you please put each new sentence on a new line in the Singularity section? 

@rainsworth I've just run this through prettier.io as per #630. Hope that's okay.

**Pull request merged**